ql/src/java/org/apache/hadoop/hive/ql/udf/ptf/ValueBoundaryScanner.java

```diff
@@ -44,10 +49,207 @@ public ValueBoundaryScanner(BoundaryDef start, BoundaryDef end, boolean nullsLas
+      rowIdx = ceilingEntry.getKey();
+    }
+
+    Object rowVal = null;
+    Object lastRowVal = null;
+
+    while (rowIdx < p.size()) {
+      rowVal = computeValue(p.getAt(rowIdx));
+      if (!isEqual(rowVal, lastRowVal)){
+        if (!cache.putIfNotFull(rowIdx, rowVal)){
+          break;
+        }
+      }
+      lastRowVal = rowVal;
+      ++rowIdx;
+
+    }
+    //Signaling end of all rows in a partition
+    if (cache.putIfNotFull(rowIdx, null)) {
+      cache.setComplete(true);
+    }
+  }
+
+  /**
+   * Uses cache content to jump backwards if possible. If not, it steps one back.
+   * @param r
+   * @param p
+   * @return pair of (row we stepped/jumped onto ; row value at this position)
+   * @throws HiveException
+   */
+  protected Pair<Integer, Object> skipOrStepBack(int r, PTFPartition p)
+          throws HiveException {
+    Object rowVal = null;
+    BoundaryCache cache = p.getBoundaryCache();
+
+    Map.Entry<Integer, Object> floorEntry = null;
+    Map.Entry<Integer, Object> ceilingEntry = null;
+
+    if (cache != null) {
+      floorEntry = cache.floorEntry(r);
+      ceilingEntry = cache.ceilingEntry(r);
+    }
+
+    if (floorEntry != null && ceilingEntry != null) {
+      r = floorEntry.getKey() - 1;
+      floorEntry = cache.floorEntry(r);
+      if (floorEntry != null) {
+        rowVal = floorEntry.getValue();
+      } else if (r >= 0){
+        rowVal = computeValue(p.getAt(r));
+      }
+    } else {
+      r--;
+      if (r >= 0) {
+        rowVal = computeValue(p.getAt(r));
+      }
+    }
+    return new ImmutablePair<>(r, rowVal);
+  }
+
+  /**
+   * Uses cache content to jump forward if possible. If not, it steps one forward.
+   * @param r
+   * @param p
+   * @return pair of (row we stepped/jumped onto ; row value at this position)
+   * @throws HiveException
+   */
+  protected Pair<Integer, Object> skipOrStepForward(int r, PTFPartition p)
+          throws HiveException {
+    Object rowVal = null;
+    BoundaryCache cache = p.getBoundaryCache();
+
+    Map.Entry<Integer, Object> floorEntry = null;
+    Map.Entry<Integer, Object> ceilingEntry = null;
+
+    if (cache != null) {
+      floorEntry = cache.floorEntry(r);
+      ceilingEntry = cache.ceilingEntry(r);
+    }
+
+    if (ceilingEntry != null && ceilingEntry.getKey().equals(r)){
+      ceilingEntry = cache.ceilingEntry(r + 1);
+    }
+    if (floorEntry != null && ceilingEntry != null) {
+      r = ceilingEntry.getKey();
+      rowVal = ceilingEntry.getValue();
+    } else {
+      r++;
+      if (r < p.size()) {
+        rowVal = computeValue(p.getAt(r));
+      }
+    }
+    return new ImmutablePair<>(r, rowVal);
+  }
+
+  /**
+   * Uses cache to lookup row value. Computes it on the fly on cache miss.
+   * @param r
+   * @param p
+   * @return row value.
+   * @throws HiveException
+   */
+  protected Object computeValueUseCache(int r, PTFPartition p) throws HiveException {
+    BoundaryCache cache = p.getBoundaryCache();
```

Comment:
Looks like copy-pate. Might worth to put to a separate method?

Reply:
These are merely variable settings, refactoring into method won't gain us much.